CRM/Core/BAO/MessageTemplate.php

```diff
@@ -687,37 +689,31 @@ protected static function parseThroughSmarty(array $mailContent, $tplParams): ar
   /**
    * Render the message template, resolving tokens and smarty tokens.
    *
+   * As with all BAO methods this should not be called directly outside
+   * of tested core code and is highly likely to change.
+   *
    * @param array $mailContent
    * @param bool $disableSmarty
-   * @param int $contactID
+   * @param int|NULL $contactID
    * @param array $smartyAssigns
    *
    * @return array
-   * @throws \CRM_Core_Exception
    */
-  protected static function renderMessageTemplate(array $mailContent, $disableSmarty, $contactID, $smartyAssigns): array {
-    $tokens = self::getTokensToResolve($mailContent);
-
-    // When using Smarty we need to pass the $escapeSmarty parameter.
-    $escapeSmarty = !$disableSmarty;
-
-    $mailContent = self::resolveDomainTokens($mailContent, $tokens, $escapeSmarty);
-
-    if ($contactID) {
-      $mailContent = self::resolveContactTokens($contactID, $tokens, $mailContent, $escapeSmarty);
-    }
-
-    // Normally Smarty is run, but it can be disabled using the disableSmarty
-    // parameter, which may be useful for non-core uses of MessageTemplate.send
-    // In particular it helps with the mosaicomsgtpl extension.
-    if (!$disableSmarty) {
-      $mailContent = self::parseThroughSmarty($mailContent, $smartyAssigns);
-    }
-    else {
-      // Since we're not relying on Smarty for this function, we DIY.
-      // strip whitespace from ends and turn into a single line
-      $mailContent['subject'] = trim(preg_replace('/[\r\n]+/', ' ', $mailContent['subject']));
+  public static function renderMessageTemplate(array $mailContent, bool $disableSmarty, $contactID, array $smartyAssigns): array {
+    CRM_Core_Smarty::singleton()->pushScope($smartyAssigns);
+    $tokenProcessor = new TokenProcessor(\Civi::dispatcher(), ['smarty' => !$disableSmarty]);
```

Comment:
This should define 'schema' => ['contactId'] in the params

Reply:
@mattwire do you want to put up a PR against the docs explaining why, what & where schema makes sense? Once that's merged I'll change this -https://docs.civicrm.org/dev/en/latest/framework/token/#token-processor ?

Reply:
(assuming we don't find the differences in the PR template to be unresolvable issues)

Reply:
(marked this as resolved as it can follow along after the docs are updated / agreed as a separate change)